### ISSUE TITLE
Unable to proxy postgres traffic with TCPIngress

### ISSUE THREAD

**Issue by landorg** [Kong/kong#7073]
### Summary
I setup a TCPIngress to forward postgres traffic (setup tls on the postgres directly as its using some application level tls) and I'm getting this error:
```
2021/04/21 13:37:43 [error] 27#0: *1002863 lua entry thread aborted: runtime error: /usr/local/share/lua/5.1/kong/init.lua:657: body must be a nil or a string
stack traceback:
coroutine 0:
	[C]: in function 'error'
	/usr/local/share/lua/5.1/kong/pdk/response.lua:849: in function 'after'
	/usr/local/share/lua/5.1/kong/init.lua:657: in function 'preread'
	preread_by_lua(nginx-kong-stream.conf:94):2: in main chunk while prereading client data, client: x.x.x.x, server: 0.0.0.0:9000
x.x.x.x [21/Apr/2021:13:37:43 +0000] TCP 500 0 0 0.017
```

### Steps To Reproduce
```yaml
apiVersion: configuration.konghq.com/v1beta1
kind: TCPIngress
metadata:
  annotations:
    kubernetes.io/ingress.class: kong
  name: terraform-state
  namespace: terraform
spec:
  rules:
  - backend:
      serviceName: terraform-postgres-cluster.terraform.svc.cluster.local
      servicePort: 5432
    port: 9000
```
values.yml
```yaml
...
  stream:
  - containerPort: 9000
    servicePort: 9000
    hostPort: 9000
```

1. `psql -h y.y.y.y -p 9000 -U postgres`

**Comment by rainest** [comment]
The stream exit handler added in https://github.com/Kong/kong/commit/13096df394e7bf9f0ba9663c40184d2a55f238f9#diff-20c108ca826c4c19db77df521ec350b0651f8c4ba0ea40511df7f9713a376e7bR811-R813 seems at odds with https://github.com/Kong/kong/blob/2.4.0/kong/runloop/handler.lua#L1090-L1096 and https://github.com/Kong/kong/blob/2.4.0/kong/tools/utils.lua#L1123-L1125, which doesn't distinguish by subsystem and always returns a table.

That said, that discrepancy just masks the attempt to send an error response of some sort. How did we intend the stream subsystem to behave in the event of a balancer error? TCP doesn't have a defined way to return errors, since it's just transport, and ICMP isn't exactly appropriate--an upstream unreachable host or similar isn't something you proxy downstream, since that's not actually the host the downstream client is trying to talk to.

**Comment by bungle** [comment]
@rainest, yes we have been thinking about that and @kikito is working on proposal I think. Yes, it is definitely hard to answer the questions, but I guess we can only log and close.

**Comment by locao** [comment]
Hi @landorg! We were unable to reproduce your actual failure with the stream subsystem. Anyway, after your report, we were able to fix the issue with the response we send on errors. With this, you'll receive a valid error message. It will be easier to identify what is the problem in your environment.

The fix can be found in #7082, which is already merged. It will be included in the next patch release.

I'm closing this for now, please feel free to reopen it if the issue persists.
